### PR TITLE
Fix variable assignment to not overwrite readonly/constant variables

### LINKED ISSUE #2324: "[datetime]$HOME=1" overwrites readonly PS variables
Steps to reproduce
------------------
```powershell
[datetime]$HOME=1
```

Expected behavior
-----------------
Throw error about `Cannot overwrite variable HOME because it is read-only or constant.`

Actual behavior
---------------
```powershell
PS D:\> [datetime]$HOME=1
PS D:\> $HOME

Monday, January 1, 0001 12:00:00 AM
```

Environment data
----------------
```powershell
PS D:\> $PSVersionTable

Name                           Value
----                           -----
PSRemotingProtocolVersion      2.3
CLRVersion
PSVersion                      6.0.0-alpha
PSEdition                      Core
GitCommitId                    v6.0.0-alpha.10-10-g87f6e8d04086da21093243acf9c6e11998e02ac8-dirty
WSManStackVersion              3.0
BuildVersion                   3.0.0.0
SerializationVersion           1.1.0.1
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
```

### PR DESCRIPTION
**Pull request by daxian-dbw**
Fix #2324

### CODE REVIEW

File: src/System.Management.Automation/engine/ShellVariable.cs [hunk 1]
```diff
@@ -668,7 +668,7 @@ private void SetValue(object value)
         {
             // Check to see if the variable is writable
 
-            if ((_options & (ScopedItemOptions.ReadOnly | ScopedItemOptions.Constant)) != ScopedItemOptions.None)
+            if (IsReadOnly || IsConstant)
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/ShellVariable.cs]: Your change is slightly more readable, but this code could be on a hot path (setting variables) so I don't see a good reason to change it.

File: test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1 [hunk 2]
```diff
@@ -140,10 +140,26 @@ Describe "Assign automatic variables" -Tags "CI" {
         & { [string]$PSScriptRoot = 'abc'; $PSScriptRoot } | Should Be abc
         & { [string]$PSCommandPath = 'abc'; $PSCommandPath } | Should Be abc
     }
+}
+
+Describe "Assign readonly/constant variables" -Tags "CI" {
+
+    It "Assign readonly/constant variables w/o type constraint" {
```
Comment (lzybkr) [inline test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1]: I think these tests are a good use case for the `-TestCases` parameter to `It`.

File: src/System.Management.Automation/engine/ShellVariable.cs (hunk 1 above)
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/ShellVariable.cs]: Yes, I see -- bytewise operation vs. possibly 2 method calls. I will revert this change.

File: test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1 (hunk 2 above)
Comment (daxian-dbw) [inline test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1]: Yep, I will make the change.

**Review by lzybkr — APPROVED** [review]

**Pull request merged**